main/lsp/requests/completion.cc

```diff
@@ -7,50 +7,122 @@ using namespace std;
 
 namespace sorbet::realmain::lsp {
 
-UnorderedMap<core::NameRef, vector<core::SymbolRef>>
-mergeMaps(UnorderedMap<core::NameRef, vector<core::SymbolRef>> &&first,
-          UnorderedMap<core::NameRef, vector<core::SymbolRef>> &&second) {
-    for (auto &other : second) {
-        first[other.first].insert(first[other.first].end(), make_move_iterator(other.second.begin()),
-                                  make_move_iterator(other.second.end()));
+namespace {
+
+core::NameRef originalName(const core::GlobalState &gs, core::NameRef name) {
+    switch (name.data(gs)->kind) {
+        case core::NameKind::UTF8:
+            return name;
+        case core::NameKind::UNIQUE:
+            return originalName(gs, name.data(gs)->unique.original);
+        case core::NameKind::CONSTANT:
+            return originalName(gs, name.data(gs)->cnst.original);
     }
-    return std::move(first);
+}
+
+vector<core::SymbolRef> ancestorsImpl(const core::GlobalState &gs, core::SymbolRef sym, vector<core::SymbolRef> &&acc) {
+    // The implementation here is similar to Symbols::derivesFrom.
+    ENFORCE(sym.data(gs)->isClassOrModuleLinearizationComputed());
+    acc.emplace_back(sym);
+
+    for (auto mixin : sym.data(gs)->mixins()) {
+        acc.emplace_back(mixin);
+    }
+
+    if (sym.data(gs)->superClass().exists()) {
+        return ancestorsImpl(gs, sym.data(gs)->superClass(), move(acc));
+    } else {
+        return move(acc);
+    }
+}
+
+// Basically the same as Module#ancestors from Ruby--but don't depend on it being exactly equal.
+// For us, it's just something that's vaguely ordered from "most specific" to "least specific" ancestor.
+vector<core::SymbolRef> ancestors(const core::GlobalState &gs, core::SymbolRef receiver) {
+    return ancestorsImpl(gs, receiver, vector<core::SymbolRef>{});
+}
+
+struct SimilarMethod final {
+    int depth;
```

Comment:
Can these be `const` if they are set once at construction?

Reply:
Making them const shouted tons of errors about implicitly deleted copy constructors. I can really look into those if you want, but I think that we're not going to get a ton of value from making these const given that its use is localized to this file.

Reply:
Ah OK, ignore my comment.